### PR TITLE
Cleanup filter pushdown logic, add support for filtering files

### PR DESCRIPTION
**Pull request by exyi**
I promised this to someone and they want to prohibit me writing new features :sweat_smile: Maybe we could claim this is just a performance improvement, anyway :thinking: 

The filter pushdown logic has grown to be a mess, I think I cleaned it up a bit by moving lot of the shared code into PushdownUtils.

### CODE REVIEW

File: src/main/scala/cognite/spark/v1/DataSetsRelation.scala [hunk 1]
```diff
@@ -24,11 +26,22 @@ class DataSetsRelation(config: RelationConfig)(val sqlContext: SQLContext)
...
   override def uniqueId(a: DataSet): String = a.id.toString
 
-  override def getStreams(filters: Array[Filter])(
+  override def getStreams(sparkFilters: Array[Filter])(
       client: GenericClient[IO],
       limit: Option[Int],
-      numPartitions: Int): Seq[fs2.Stream[IO, DataSet]] =
-    Seq(client.dataSets.list(limit))
+      numPartitions: Int): Seq[fs2.Stream[IO, DataSet]] = {
+    // not super necessary to do filter pushdown here, but it looks nicer when we don't fetch all datasets in Jetfire
```
Comment (hakontro) [inline src/main/scala/cognite/spark/v1/DataSetsRelation.scala:33]: Maybe obvious what this means, but I didn't get it.. can you elaborate/rephrase?

File: src/main/scala/cognite/spark/v1/PushdownUtilities.scala [hunk 2]
```diff
@@ -24,6 +31,28 @@ object PushdownUtilities {
   def pushdownToUri(parameters: Seq[Map[String, String]], uri: Uri): Seq[Uri] =
     parameters.map(params => uri.params(params))
 
+  def pushdownToFilters[F](
+      sparkFilters: Array[Filter],
+      mapping: Map[String, String] => F,
+      allFilter: F): (Vector[CogniteId], Vector[F]) = {
```
Comment (hakontro) [inline src/main/scala/cognite/spark/v1/PushdownUtilities.scala:37]: Can you think of a better name than `allFilter`? (I couldn't)

File: src/main/scala/cognite/spark/v1/PushdownUtilities.scala [hunk 3]
```diff
@@ -198,21 +230,93 @@ object PushdownUtilities {
...
+  def getFromIds[T, F[_]: Concurrent](
+      ids: Vector[CogniteId],
+      resource: RetrieveByIdsWithIgnoreUnknownIds[T, F] with RetrieveByExternalIdsWithIgnoreUnknownIds[
+        T,
+        F]
+  ): Stream[F, T] =
+    getFromIds(
+      ids,
+      resource.retrieveByIds(_, ignoreUnknownIds = true),
+      resource.retrieveByExternalIds(_, ignoreUnknownIds = true)
```
Comment (hakontro) [inline src/main/scala/cognite/spark/v1/PushdownUtilities.scala:263]: Scala really is the best SDK

**Review by hakontro — CHANGES_REQUESTED** [review]
LGTM! Just needs to add new entries for `_UpdatedTime` to the pushdowncolumns table in the Readme.
We have feature freeze from today, and while I agree that it doesn't really add new features, I think it has a risk of introducing some new bugs.. so let's keep it in the Spark Datasource at least, but releasing a new version is fine.

File: src/main/scala/cognite/spark/v1/PushdownUtilities.scala (hunk 2 above)
Comment (vuhainguyen) [reply src/main/scala/cognite/spark/v1/PushdownUtilities.scala:37]: Stanislav in the next 3 days 😆 
image
Just kidding

File: src/main/scala/cognite/spark/v1/DataSetsRelation.scala (hunk 1 above)
Comment (exyi) [reply src/main/scala/cognite/spark/v1/DataSetsRelation.scala:33]: sorry it's probably just confusing. I was just justifying doing filter pushdown for datasets when you are quite unlikely to have many datasets in a project.

**Comment by exyi** [comment]
I have added better documentation for the filter pushdown, do you like it this way @hakontro?

**Review by hakontro — APPROVED** [review]
Good stuff!

**Pull request merged**